modules/bls-dkg/index.js

```diff
@@ -43,6 +44,75 @@ function bigIntToBuffer(bigint) {
   return hexToBuffer(bigint.toString(16))
 }
 
+/** Integer to Octet Stream */
+function i2osp(value, length) {
+  if (value < 0 || value >= 1n << BigInt(8 * length)) {
+    throw new Error(`bad I2OSP call: value=${value} length=${length}`);
+  }
+  const res = Array.from({ length }).fill(0);
+  for (let i = length - 1; i >= 0; i--) {
+    res[i] = value & 0xff;
+    value >>>= 8;
+  }
+  return new Uint8Array(res);
+}
+
+/** Concatenate multiple buffers into one */
+function concatBytes(...arrays) {
+  if (arrays.length === 1) {
+    return arrays[0];
+  }
+  const length = arrays.reduce((a, arr) => a + arr.length, 0);
+  const result = new Uint8Array(length);
+  for (let i = 0, pad = 0; i < arrays.length; i++) {
+    const arr = arrays[i];
+    result.set(arr, pad);
+    pad += arr.length;
+  }
+  return result;
+}
+
+/** Get the indices from a sub key path string */
+function getSubKeyPathIndices(path) {
+  return path.replace(/^m\//, '')
+    .split('/')
+    .map((index) => parseInt(index, 10));
+}
+
+/** Calculate offset from a public key */
+function calculateOffset(pk, index) {
+  const salt = i2osp(index, 4);
+  const buffer = concatBytes(pk, salt);
+  const digest = createHash('sha256').update(buffer).digest();
+  return bufferToBigInt(digest);
+}
+
+/** Derive a public key */
+function publicDerive(pk, path) {
+  const indices = getSubKeyPathIndices(path);
+  function deriveIndex(pk, index) {
+    const pkBuf = bigIntToBuffer(pk);
+    const offset = calculateOffset(pkBuf, index);
+    const offsetMod = new math.Fq(offset);
+    const pkPoint = bls.PointG1.fromCompressedHex(pkBuf);
+    const publicChild = pkPoint.add(bls.PointG1.BASE.multiply(offsetMod));
+    return bufferToBigInt(publicChild.toCompressedHex());
+  }
+  return indices.reduce(deriveIndex, pk);
+}
+
+/** Derive a private key */
+function privateDerive(parentSk, path) {
```

Comment:
will we have the parentSk ever aggregated in one place?

Reply:
No, it is distributed through the secret shares to the participants.